File: IMS/Models/MaterialLineDisplay.cs
using System.Globalization;
using System.Text;

namespace IMS.Models;

internal static class MaterialLineDisplay
{
    public static string FormatStage(string? stage)
    {
        if (string.IsNullOrWhiteSpace(stage))
            return "—";
        return CultureInfo.CurrentCulture.TextInfo.ToTitleCase(stage.Replace('_', ' '));
    }

    public static string FormatAssignment(string? assignmentType)
    {
        if (string.IsNullOrWhiteSpace(assignmentType))
            return "BOM";
        return assignmentType.ToUpperInvariant() switch
        {
            "BOM" => "BOM",
            "MANUAL" => "Manual",
            "OVERRIDE" => "Override",
            _ => assignmentType
        };
    }

    public static string FormatLastEvent(IReadOnlyList<MaterialStageEvent> events)
    {
        if (events.Count == 0)
            return "—";
        var e = events[^1];
        var when = e.At?.ToString("g", CultureInfo.CurrentCulture) ?? "";
        var place = string.IsNullOrWhiteSpace(e.Godown) ? "" : $" @ {e.Godown}";
        return $"{FormatStage(e.Stage)}{place} · {when}".Trim(' ', '·');
    }

    public static string FormatHistoryToolTip(IReadOnlyList<MaterialStageEvent> events)
    {
        if (events.Count == 0)
            return "No stage history yet.";
        var sb = new StringBuilder();
        foreach (var e in events)
        {
            if (sb.Length > 0)
                sb.AppendLine();
            var when = e.At?.ToString("g", CultureInfo.CurrentCulture) ?? "—";
            var who = string.IsNullOrWhiteSpace(e.By) ? "" : $" · {e.By}";
            var qty = e.Qty > 0 ? $" · qty {e.Qty:N2}" : "";
            var note = string.IsNullOrWhiteSpace(e.Note) ? "" : $" — {e.Note}";
            sb.Append($"{FormatStage(e.Stage)} · {when}{who}{qty}{note}");
        }
        return sb.ToString();
    }
}

File: ims-web/src/work-order/materialLineDisplay.ts
import type { MaterialStageEvent } from '../api/productionOrders';

export function formatMaterialStage(stage?: string): string {
  if (!stage?.trim()) return '—';
  return stage
    .replace(/_/g, ' ')
    .replace(/\b\w/g, (char) => char.toUpperCase());
}

export function formatMaterialAssignment(assignmentType?: string): string {
  if (!assignmentType?.trim()) return 'BOM';
  switch (assignmentType.toUpperCase()) {
    case 'BOM':
      return 'BOM';
    case 'MANUAL':
      return 'Manual';
    case 'OVERRIDE':
      return 'Override';
    default:
      return assignmentType;
  }
}

export function formatLastMaterialEvent(events?: MaterialStageEvent[]): string {
  if (!events?.length) return '—';
  const event = events[events.length - 1];
  const when = event.at ? new Date(event.at).toLocaleString() : '';
  const place = event.godown?.trim() ? ` @ ${event.godown}` : '';
  const summary = `${formatMaterialStage(event.stage)}${place}${when ? ` · ${when}` : ''}`.trim();
  return summary.replace(/\s·\s$/, '') || '—';
}

export function formatMaterialHistoryTooltip(events?: MaterialStageEvent[]): string {
  if (!events?.length) return 'No stage history yet.';
  return events
    .map((event) => {
      const when = event.at ? new Date(event.at).toLocaleString() : '—';
      const who = event.by?.trim() ? ` · ${event.by}` : '';
      const qty = Number(event.qty) > 0 ? ` · qty ${Number(event.qty).toFixed(2)}` : '';
      const note = event.note?.trim() ? ` — ${event.note}` : '';
      return `${formatMaterialStage(event.stage)} · ${when}${who}${qty}${note}`;
    })
    .join('\n');
}
